styx-scheduler-service/src/test/java/com/spotify/styx/state/handlers/TerminationHandlerTest.java

```diff
@@ -126,13 +152,114 @@ public void shouldScheduleRetryOf10MinutesOnMissingDependencies() {
   }
 
   @Test
-  public void shouldFailOnFailFastExitCodeReceived() {
-    StateData data = data(1, 1.0, Optional.of(50));
-    RunState maxedTerm = RunState.create(WORKFLOW_INSTANCE, FAILED, data, NOW, COUNTER);
-    terminationHandler.transitionInto(maxedTerm);
+  public void shouldStopOnFailFastExitCodeReceived() {
+    var data = data(1, 1.0, Optional.of(50));
+    var failed = RunState.create(WORKFLOW_INSTANCE, FAILED, data, NOW, COUNTER);
+    terminationHandler.transitionInto(failed);
+    verify(stateManager).receiveIgnoreClosed(Event.stop(WORKFLOW_INSTANCE), COUNTER);
+  }
+
+  @Test
+  public void shouldStopOnWorkflowNotFound() {
+    var data = data(1, 1.0, Optional.empty());
+    var failed = RunState.create(WORKFLOW_INSTANCE, FAILED, data, NOW, COUNTER);
+    when(workflows.get()).thenReturn(Map.of());
+    terminationHandler.transitionInto(failed);
+    verify(stateManager).receiveIgnoreClosed(Event.stop(WORKFLOW_INSTANCE), COUNTER);
+  }
+
+  @Test
+  public void shouldScheduleRetryOnRetryConditionMet() {
+    var data = data(1, 1.0, Optional.of(1));
+    var nonZeroTerm = RunState.create(WORKFLOW_INSTANCE, TERMINATED, data, NOW, COUNTER);
+    Duration expectedDelay = Duration.ofMillis(4711);
+    when(retryUtil.calculateDelay(anyInt())).thenReturn(expectedDelay);
+    when(workflows.get()).thenReturn(Map.of(WORKFLOW_WITH_RETRY_CONDITION.id(), WORKFLOW_WITH_RETRY_CONDITION));
+    terminationHandler.transitionInto(nonZeroTerm);
+    verify(stateManager).receiveIgnoreClosed(Event.retryAfter(WORKFLOW_INSTANCE, expectedDelay.toMillis()), COUNTER);
+  }
+
+  @Test
+  public void shouldStopOnRetryConditionNotMet() {
+    var data = data(1, 1.0, Optional.of(2));
+    var nonZeroTerm = RunState.create(WORKFLOW_INSTANCE, TERMINATED, data, NOW, COUNTER);
+    when(workflows.get()).thenReturn(Map.of(WORKFLOW_WITH_RETRY_CONDITION.id(), WORKFLOW_WITH_RETRY_CONDITION));
+    terminationHandler.transitionInto(nonZeroTerm);
     verify(stateManager).receiveIgnoreClosed(Event.stop(WORKFLOW_INSTANCE), COUNTER);
   }
 
+  @Test
+  public void shouldEvaluateToTrue() {
```

Comment:
What is this testing? all threads seem to be identical.

Reply:
This is to ensure thread-safety of the implementation because `expressionParser` is shared by threads.

Reply:
It also serves as performance test. Maybe I should put a timeout.

Reply:
Ok, let's reflect that in the test name then?

Reply:
Yeah, absolutely.